internal/v4/api_test.go

```diff
@@ -885,6 +917,85 @@ func (s *APISuite) TestMetaPerm(c *gc.C) {
 			Write: []string{},
 		},
 	})
+
+	// Restore some write rights to the stable channel.
+	s.assertPutAsAdmin(c, "trusty/wordpress-1/meta/perm/write", []string{"bob"})
+
+	// ~charmers/trusty/wordpress-1 has been published only to the
+	// stable channel. If we specify a channel in a perm put
+	// request, we'll get an error because the channel isn't valid for
+	// that entity.
+	s.doAsUser("charmers", func() {
+		httptesting.AssertJSONCall(c, httptesting.JSONCallParams{
+			Handler: s.srv,
+			Do:      bakeryDo(nil),
+			Method:  "PUT",
+			JSONBody: params.PermRequest{
+				Read:  []string{"foo"},
+				Write: []string{"bar"},
+			},
+			URL:          storeURL("trusty/wordpress-1/meta/perm?channel=development"),
+			ExpectStatus: http.StatusNotFound,
+			ExpectBody: params.Error{
+				Code:    params.ErrNotFound,
+				Message: `cs:trusty/wordpress-1 not found in development channel`,
+			},
+		})
+	})
+
+	// Similarly, we should be able to specify a channel on read
+	// to read a different channel.
+	s.doAsUser("bob", func() {
+		s.assertGet(c, "trusty/wordpress/meta/perm?channel=unpublished", params.PermResponse{
+			Read:  []string{"bob"},
+			Write: []string{"admin"},
+		})
+		s.assertGet(c, "wordpress/meta/perm?channel=development", params.PermResponse{
+			Read:  []string{"bob", "charlie"},
+			Write: []string{"charmers"},
+		})
+	})
+
+	// We can't write to a channel that the charm's not in.
```

Comment:
How would you propose changing it?
Currently we have a simple model that resolves entities before we get to any entity-specific endpoint. That's nice and easy to explain - all endpoints within an entity are treated the same.

To do it differently, you'd need to avoid resolving until the actual endpoint is encountered, but that makes bulk updates much harder (what do you do for a PUT to wordpress-1/meta/any which includes both extra-info and perm?) and everything more complex AFAICS, for not *that* much gain.

By making the security decision in one place rather than in each individual handler, we gain more confidence in our security. I don't think that this is worth compromising that for.

Another approach that's compatible with the current scheme would be to have new meta endpoints that are specifically oriented to changing channel-specific ACLs, for example meta/perm.stable, meta/perm.development, etc.